skbio/diversity/_driver.py

```diff
@@ -0,0 +1,304 @@
+        error will depend on what was invalid.
+    TypeError
+        If invalid method-specific parameters are provided.
+
+    See Also
+    --------
+    unweighted_unifrac
+    weighted_unifrac
+    scipy.spatial.distance.pdist
+    skbio.diversity.alpha_diversity
+
+    Notes
+    -----
+    The value that you provide for for ``metric`` can be either a string (e.g.,
+    ``"unweighted_unifrac"``) or a function
+    (e.g., ``skbio.diversity.beta.unweighted_unifrac``). The metric should
+    generally be passed as a string, as this often uses an optimized version
+    of the metric. For example, passing  ``"unweighted_unifrac"`` (a string)
+    will be hundreds of times faster than passing the function
+    ``skbio.diversity.beta.unweighted_unifrac``. The latter is faster if
+    computing only one or a few distances, but in these cases the difference in
+    runtime is negligible, so it's safer to just err on the side of passing
+    ``metric`` as a string.
+
+    Validation of input data confirms the following:
+     * ``counts`` data can be safely cast to integers
+     * there are no negative values in ``counts``
+     * ``counts`` has the correct number of dimensions
+     * all vectors in ``counts`` are of equal length
+     * the correct number of ``ids`` is provided (if any are provided)
+    And for phylogenetic diversity metrics:
+     * ``otu_ids`` does not contain duplicate values
+     * the length of each ``counts`` vector is equal to ``len(otu_ids)``
+     * ``tree`` is rooted
+     * ``tree`` has more than one node
+     * all nodes in ``tree`` except for the root node have branch lengths
+     * all tip names in ``tree`` are unique
+     * all ``otu_ids`` correspond to tip names in ``tree``
+
+    """
+    if validate:
+        counts = _validate_counts_matrix(counts, ids=ids)
+
+    if metric == 'unweighted_unifrac':
+        otu_ids, tree, kwargs = _get_phylogenetic_kwargs(counts, **kwargs)
+        metric, counts_by_node = _setup_multiple_unweighted_unifrac(
+                counts, otu_ids=otu_ids, tree=tree, validate=validate)
+        counts = counts_by_node
+    elif metric == 'weighted_unifrac':
+        # get the value for normalized. if it was not provided, it will fall
+        # back to the default value inside of _weighted_unifrac_pdist_f
+        normalized = kwargs.pop('normalized',
+                                _normalize_weighted_unifrac_by_default)
+        otu_ids, tree, kwargs = _get_phylogenetic_kwargs(counts, **kwargs)
+        metric, counts_by_node = _setup_multiple_weighted_unifrac(
+                counts, otu_ids=otu_ids, tree=tree, normalized=normalized,
+                validate=validate)
+        counts = counts_by_node
+    elif callable(metric):
+        metric = partial(metric, **kwargs)
+        # remove all values from kwargs, since they have already been provided
+        # through the partial
+        kwargs = {}
```

Comment:
Scratch that, it works for scipy `pdist` metrics passed as `str`. I bet `pdist` kwargs don't apply to `callable` metrics (I don't know how they could get that to work)